lib/collection/src/collection_manager/optimizers/segment_optimizer.rs

```diff
@@ -422,20 +430,32 @@ pub trait SegmentOptimizer {
             })
             .collect();
 
-        // TODO: use a different approach to select the key to use for defragmentation
-        for segment in &segments {
-            let defragment_key = segment
-                .read()
-                .payload_index
-                .borrow()
-                .field_indexes
-                .iter()
-                .next()
-                .map(|(k, _)| k.clone());
-
-            if let Some(defragment_key) = defragment_key {
-                segment_builder.set_defragment_key(defragment_key);
-                break;
+        if let Some(defragmentation_config) = self.defragmentation_config() {
+            segment_builder.set_defragment_keys(
+                defragmentation_config
+                    .keys
+                    .iter()
+                    .filter_map(|i| PayloadKeyType::from_str(i).ok())
+                    .collect(),
+            );
+        } else {
+            // Automatically select the defragmentation key.
+            //
+            // TODO: try to make this more intelligent
+            for segment in &segments {
+                let defragment_key = segment
+                    .read()
+                    .payload_index
+                    .borrow()
+                    .field_indexes
+                    .iter()
+                    .next()
+                    .map(|(k, _)| k.clone());
+
+                if let Some(defragment_key) = defragment_key {
+                    segment_builder.set_defragment_keys(vec![defragment_key]);
+                    break;
+                }
             }
```

Comment:
I like that we have an automatic function, assuming we can come up with a good default.

However, I wonder if we also want to be able to fully disable it. I'm not sure yet.

We can possibly do this with `null`, `[]`, `["some", "keys]` as defragmentation keys. Where an empty list would mean no defragging at all.

Reply:
I had the same idea but it seems like protobuf doesn't support optional+repeated fields, so we can't differentiate between `null` and `[]`, at least not without a workaround (please correct me if I'm wrong).
Also I can imagine this becoming a bit confusing, because `null` can also mean "no defragmentation" to some users.
Lastly I couldn't think of a reason why we want to disable it, especially because we don't enforce any order currently.

This were my thoughts when implementing, what do you think?

Reply:
> (...) so we can't differentiate between null and [] (...)

Yeah we have a similar-ish problem in Python actually. There we cannot distinguish between a field being `None` or non-existent.

> Lastly I couldn't think of a reason why we want to disable it, especially because we don't enforce any order currently.

If defragmentation doesn't work for some use case, it only adds overhead during optimization time.

Reply:
> If defragmentation doesn't work for some use case, it only adds overhead during optimization time.

You're right, but looking at the execution metrics I figured we can trade between complexity and a few ms that it takes longer for sorting, even if it is not needed. Or do you disagree?